File: src/main/java/com/teamwizardry/librarianlib/ragdoll/line/Link2D.java
package com.teamwizardry.librarianlib.ragdoll.line;

import com.teamwizardry.librarianlib.math.Vec2d;

/**
 * Created by Saad on 12/7/2016.
 */
public class Link2D {

    public PointMass2D a, b;
    public float restingDistance;
    public float strength;
    public boolean pinA, pinB;
    
    public Link2D(PointMass2D a, PointMass2D b, float strength) {
        this.a = a;
        this.b = b;
        this.strength = strength;
        this.restingDistance = (float) a.pos.sub(b.pos).length();
    }

    public Link2D(PointMass2D a, PointMass2D b, float restingDistance, float strength) {
        this.a = a;
        this.b = b;
        this.strength = strength;
        this.restingDistance = restingDistance;
    }

    public Link2D pinA() { pinA = true; return this; }
    public Link2D pinB() { pinB = true; return this; }
    
    public void resolve() {
        if (a.pin && b.pin) return;

        Vec2d posDiff = a.pos.sub(b.pos);
        double d = posDiff.length();

        double difference = (restingDistance - d) / d;
        difference *= strength;
        if (!( a.pin || pinA ) && !( b.pin || pinB )) // neither are pinned
            difference = difference / 2.0;

        Vec2d translate = posDiff.mul(difference);

        if (!( a.pin || pinA )) a.pos = a.pos.add(translate);

        if (!( b.pin || pinB )) b.pos = b.pos.sub(translate);
    }
}

File: src/main/java/com/teamwizardry/librarianlib/bloat/ragdoll/line/Link2D.kt
package com.teamwizardry.librarianlib.bloat.ragdoll.line

/**
 * Created by Saad on 12/7/2016.
 */
class Link2D {

    var a: PointMass2D
    var b: PointMass2D
    var restingDistance: Float = 0.toFloat()
    var strength: Float = 0.toFloat()
    var pinA: Boolean = false
    var pinB: Boolean = false

    constructor(a: PointMass2D, b: PointMass2D, strength: Float) {
        this.a = a
        this.b = b
        this.strength = strength
        this.restingDistance = a.pos.sub(b.pos).length().toFloat()
    }

    constructor(a: PointMass2D, b: PointMass2D, restingDistance: Float, strength: Float) {
        this.a = a
        this.b = b
        this.strength = strength
        this.restingDistance = restingDistance
    }

    fun pinA(): Link2D {
        pinA = true
        return this
    }

    fun pinB(): Link2D {
        pinB = true
        return this
    }

    fun resolve() {
        if (a.pin && b.pin) return

        val posDiff = a.pos.sub(b.pos)
        val d = posDiff.length()

        var difference = (restingDistance - d) / d
        difference *= strength.toDouble()
        if (!(a.pin || pinA) && !(b.pin || pinB))
        // neither are pinned
            difference = difference / 2.0

        val translate = posDiff.mul(difference)

        if (!(a.pin || pinA)) a.pos = a.pos.add(translate)

        if (!(b.pin || pinB)) b.pos = b.pos.sub(translate)
    }
}
